crates/re_renderer/src/mesh.rs

```diff
@@ -79,12 +84,87 @@ pub mod mesh_vertices {
 pub struct Mesh {
     pub label: DebugLabel,
 
+    /// Length must be a multipel of three.
     pub indices: Vec<u32>, // TODO(andreas): different index formats?
+
     pub vertex_positions: Vec<glam::Vec3>,
-    pub vertex_data: Vec<mesh_vertices::MeshVertexData>,
+
+    /// Per-vertex albedo color with unmulitplied/separate alpha.
+    /// Must be equal in length to [`Self::vertex_positions`].
+    pub vertex_colors: Vec<[u8; 4]>,
+
+    /// Must be equal in length to [`Self::vertex_positions`].
+    /// Use ZERO for unshaded.
+    pub vertex_normals: Vec<glam::Vec3>,
+
+    /// Must be equal in length to [`Self::vertex_positions`].
+    pub vertex_texcoords: Vec<glam::Vec2>,
+
     pub materials: SmallVec<[Material; 1]>,
 }
 
+impl Mesh {
+    pub fn sanity_check(&self) -> Result<(), MeshError> {
+        crate::profile_function!();
+
+        let num_pos = self.vertex_positions.len();
+        let num_color = self.vertex_colors.len();
+        let num_normals = self.vertex_normals.len();
+        let num_texcoords = self.vertex_texcoords.len();
+
+        if num_pos != num_color {
+            return Err(MeshError::WrongNumberOfColors { num_pos, num_color });
+        }
+        if num_pos != num_normals {
+            return Err(MeshError::WrongNumberOfNormals {
+                num_pos,
+                num_normals,
+            });
+        }
+        if num_pos != num_texcoords {
+            return Err(MeshError::WrongNumberOfTexcoord {
+                num_pos,
+                num_texcoords,
+            });
+        }
+
+        if self.indices.len() % 3 != 0 {
+            return Err(MeshError::WrongNumberOfIndices {
+                num_indices: self.indices.len(),
+            });
+        }
+
+        for &index in &self.indices {
+            if num_pos <= index as usize {
+                return Err(MeshError::IndexOutOfBounds { num_pos, index });
+            }
+        }
```

Comment:
I really like having these here, but makes me wonder if we can maybe use this directly in the mesh3d.rs component type 🤔. Right now that would ofc create unwanted dependencies, so probably not worth it.

Reply:
Yeah, I would like to merge them at some point